src/devices/Gpio/Drivers/VirtualGpioController.cs

```diff
@@ -0,0 +1,318 @@
+            }
+
+            return _pins[pinNumber].GetPinMode();
+        }
+
+        /// <inheritdoc/>
+        public override bool IsPinModeSupported(int pinNumber, PinMode mode)
+        {
+            return _pins[pinNumber].IsPinModeSupported(mode);
+        }
+
+        /// <summary>
+        /// Checks if a specific pin is open.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        /// <returns>The status if the pin is open or closed.</returns>
+        public new bool IsPinOpen(int pinNumber)
+        {
+            return _openPins.ContainsKey(pinNumber);
+        }
+
+        /// <summary>
+        /// Opens a pin and sets it to a specific mode.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        /// <param name="mode">The mode to be set.</param>
+        public new GpioPin OpenPin(int pinNumber, PinMode mode)
+        {
+            var pin = OpenPin(pinNumber);
+            _pins[pinNumber].SetPinMode(mode);
+            return pin;
+        }
+
+        /// <summary>
+        /// Opens a pin and sets it to a specific mode and value.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        /// <param name="mode">The mode to be set.</param>
+        /// <param name="initialValue">The initial value to be set if the mode is output. The driver will attempt to set the mode without causing glitches to the other value.
+        /// (if <paramref name="initialValue"/> is <see cref="PinValue.High"/>, the pin should not glitch to low during open)</param>
+        public new GpioPin OpenPin(int pinNumber, PinMode mode, PinValue initialValue)
+        {
+            var pin = OpenPin(pinNumber, mode);
+            _pins[pinNumber].Write(initialValue);
+            return pin;
+        }
+
+        /// <summary>
+        /// Opens a pin in order for it to be ready to use.
+        /// The driver attempts to open the pin without changing its mode or value.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        public new GpioPin OpenPin(int pinNumber)
+        {
+            if (IsPinOpen(pinNumber))
+            {
+                return _openPins[pinNumber];
+            }
+
+            _openPins.TryAdd(pinNumber, new VirtualGpioPin(_pins[pinNumber], pinNumber));
+            return _openPins[pinNumber];
+        }
+
+        /// <inheritdoc/>
+        public override ComponentInformation QueryComponentInformation()
+        {
+            ComponentInformation self = new ComponentInformation(this, "Virtual GPIO Controller");
+
+            // PinCount is not added on purpose, because the property throws NotSupportedException on some hardware
```

Comment:
Comment doesn't apply here, the list of available pins for the virtual controller is known.